modules/core/src/main/java/org/apache/ignite/internal/util/GridConcurrentMultiPairQueue.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.util;
+
+import org.apache.ignite.internal.util.typedef.T2;
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Concurrent queue that wraps collection of {@code Pair<K, V[]>}
+ * The only garantee {@link #poll} or {@link #next} provided is sequentially emptify values per key array.
+ * i.e. input like: <br>
+ * p1 = new Pair<1, [1, 3, 5, 7]> <br>
+ * p2 = new Pair<2, [2, 3]> <br>
+ * p3 = new Pair<3, [200, 100]> <br>
+ * and further sequence of {@code poll} or {@code forEach} calls may produce output like: <br>
+ * [3, 200], [3, 100], [1, 1], [1, 3], [1, 5], [1, 7], [2, 2], [2, 3]
+ *
+ * @param <K> The type of key in input pair collection.
+ * @param <V> The type of value array.
+ */
+public class GridConcurrentMultiPairQueue<K, V> {
+    /** */
+    public static final GridConcurrentMultiPairQueue EMPTY =
+        new GridConcurrentMultiPairQueue<>(Collections.emptyMap());
+
+    /** Inner holder. */
+    private final V[][] vals;
+
+    /** Storage for every array length. */
+    private final int[] lenSeq;
+
+    /** Current absolute position. */
+    private final AtomicInteger pos = new AtomicInteger();
+
+    /** Precalculated max position. */
+    private final int maxPos;
+
+    /** Keys array. */
+    private final K[] keysArr;
+
+    /** Thread local cached segment. */
+    private ThreadLocal<Integer> cachedSegment = ThreadLocal.withInitial(() -> 0);
```

Comment:
No one calls ThreadLocal#remove on this one, this will lead to memory leak in checkpoint thread.

Reply:
no one calls "poll" now + GridConcurrentMultiPairQueue instance has life time smaller than checkpoint frequency, do we need some changes here ?